tempesta_fw/http_parser.c

```diff
@@ -457,6 +457,134 @@ parse_int_hex(unsigned char *data, size_t len, unsigned long *acc)
 	return CSTR_POSTPONE;
 }
 
+/**
+ * Add unconditionaly hop-by-hop headers to list of hop-by-hop headers names.
+ * Do not init raw headers table until we found that Connection header contains
+ * atleast one raw header.
+ */
+static void
+__hbh_parser_init(TfwHttpHbhHdrs *hbh_hdrs, bool req)
+{
+	/*
+	 * Connection header is hop-by-hop header.
+	 *
+	 * Server header isn't defined as hop-by-hop by the RFC, but we
+	 * don't show protected server to world. User-agent and Server headers
+	 * have the same id in tfw_http_hdr_t, so Server header must be marked
+	 * only for responses.
+	 */
+	hbh_hdrs->spec |= 0x1 << TFW_HTTP_HDR_CONNECTION;
+	if (!req)
+		hbh_hdrs->spec |= 0x1 << TFW_HTTP_HDR_SERVER;
+}
+
+#define MARK_SPEC_HBH(hid)						\
+	if ((msg->parser.hbh_parser.spec & (0x1 << hid))		\
+	    && !TFW_STR_EMPTY(&msg->parser.hdr))			\
+		msg->parser.hdr.flags |= TFW_STR_HBH_HDR;
+
+/**
+ * Mark raaw header @hdr as hop-by-hop if its name was listed in Connection
+ * header
+ */
+static void
+mark_raw_hbh(TfwHttpMsg *hm, TfwStr *hdr)
+{
+	TfwHttpHdrTbl *ht = hm->parser.hbh_parser.raw;
+	unsigned int i;
+
+	if (!ht)
+		return;
+
+	for (i = 0; i < ht->off; ++i) {
```

Comment:
Does RFC allow setting headers, mentioned in Connection, before Connection header? The code is dangerous, just consider the example:

        GET / HTTP/1.1\r\n
        Connection: keep-alive, Host, User-Agent\r\n
        Host: foo.com\r\n
        User-Agent: malicious client\r\n
        \r\n

So server will receive malformed request.

We should run the loop at least from TFW_HTTP_HDR_RAW. Probably there is some limitation which headers can be mentioned in Connection header... The other way (I think the safest one) is to remove Keep-Alive only and ignore any other mentioned headers in Connection. However, we still use current mechanism for the headers removal such that we'll be able to remove other headers which are known to be mentioned in Connection. I argue against blind headers removal.